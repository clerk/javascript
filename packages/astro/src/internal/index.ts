/**
 * The following code will be used in order to be injected as script via the astro integration.
 * F.e.
 *
 * injectScript('before-hydration', `...`)
 */

import { createClerkInstance } from './create-clerk-instance';
import { createInjectionScriptRunner } from './create-injection-script-runner';

const runInjectionScript = createInjectionScriptRunner(createClerkInstance);

export { runInjectionScript };

export { generateSafeId } from './utils/generateSafeId';
