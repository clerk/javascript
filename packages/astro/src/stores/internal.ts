import type {
  ActiveSessionResource,
  Clerk,
  ClientResource,
  InitialState,
  OrganizationResource,
  UserResource,
} from '@clerk/types';
import { atom, map } from 'nanostores';

export const $csrState = map<{
  isLoaded: boolean;
  client: ClientResource | undefined | null;
  user: UserResource | undefined | null;
  session: ActiveSessionResource | undefined | null;
  organization: OrganizationResource | undefined | null;
}>({
  isLoaded: false,
  client: null,
  user: null,
  session: null,
  organization: null,
});

export const $initialState = map<InitialState>();

// Use atom instead of `map` to prohibit key changes and allow only replacing the whole object
export const $clerk = atom<Clerk | null>(null);
