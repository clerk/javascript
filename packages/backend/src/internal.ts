export { constants } from './constants';
export { createRedirect } from './createRedirect';
export type { RedirectFun } from './createRedirect';

export type { CreateAuthenticateRequestOptions } from './tokens/factory';
export { createAuthenticateRequest } from './tokens/factory';

export { debugRequestState } from './tokens/request';

export type { AuthenticateRequestOptions, OrganizationSyncOptions } from './tokens/types';

export type { SignedInAuthObjectOptions, SignedInAuthObject, SignedOutAuthObject } from './tokens/authObjects';
export { makeAuthObjectSerializable, signedOutAuthObject, signedInAuthObject } from './tokens/authObjects';

export { AuthStatus } from './tokens/authStatus';
export type { RequestState, SignedInState, SignedOutState } from './tokens/authStatus';

export { decorateObjectWithResources, stripPrivateDataFromObject } from './util/decorateObjectWithResources';

export { createClerkRequest } from './tokens/clerkRequest';
export type { ClerkRequest } from './tokens/clerkRequest';

export { reverificationError, reverificationErrorResponse } from '@clerk/shared/authorization-errors';
