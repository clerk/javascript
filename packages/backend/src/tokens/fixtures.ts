export const mockJwt =
  'eyJhbGciOiJSUzI1NiIsImtpZCI6Imluc18yR0lvUWhiVXB5MGhYN0IyY1ZrdVRNaW5Yb0QiLCJ0eXAiOiJKV1QifQ.eyJhenAiOiJodHRwczovL2FjY291bnRzLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsImV4cCI6MTY2NjY0ODMxMCwiaWF0IjoxNjY2NjQ4MjUwLCJpc3MiOiJodHRwczovL2NsZXJrLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsIm5iZiI6MTY2NjY0ODI0MCwic2lkIjoic2Vzc18yR2JEQjRlbk5kQ2E1dlMxenBDM1h6Zzl0SzkiLCJzdWIiOiJ1c2VyXzJHSXBYT0VwVnlKdzUxcmtabjlLbW5jNlN4ciJ9.n1Usc-DLDftqA0Xb-_2w8IGs4yjCmwc5RngwbSRvwevuZOIuRoeHmE2sgCdEvjfJEa7ewL6EVGVcM557TWPW--g_J1XQPwBy8tXfz7-S73CEuyRFiR97L2AHRdvRtvGtwR-o6l8aHaFxtlmfWbQXfg4kFJz2UGe9afmh3U9-f_4JOZ5fa3mI98UMy1-bo20vjXeWQ9aGrqaxHQxjnzzC-1Kpi5LdPvhQ16H0dPB8MHRTSM5TAuLKTpPV7wqixmbtcc2-0k6b9FKYZNqRVTaIyV-lifZloBvdzlfOF8nW1VVH_fx-iW5Q3hovHFcJIULHEC1kcAYTubbxzpgeVQepGg';

export const mockJwtHeader = {
  alg: 'RS256',
  kid: 'ins_2GIoQhbUpy0hX7B2cVkuTMinXoD',
  typ: 'JWT',
};

export const mockJwtPayload = {
  azp: 'https://accounts.inspired.puma-74.lcl.dev',
  exp: 1666648310,
  iat: 1666648250,
  iss: 'https://clerk.inspired.puma-74.lcl.dev',
  nbf: 1666648240,
  sid: 'sess_2GbDB4enNdCa5vS1zpC3Xzg9tK9',
  sub: 'user_2GIpXOEpVyJw51rkZn9Kmnc6Sxr',
};

export const mockRsaJwkKid = 'ins_2GIoQhbUpy0hX7B2cVkuTMinXoD';

export const mockRsaJwk = {
  use: 'sig',
  kty: 'RSA',
  kid: mockRsaJwkKid,
  alg: 'RS256',
  n: 'u0tNUitBZmcGYMWcqvaRBaJe0XmTQ738RHYoHjhYANyeOkysuu4L_Rqr-fmTXsbebrTp7_OewIqsJXImEWB_WQ3HN9lAkOMCCGDU1udsz_sl1Kwy5JZ7x8Nr4ghXJagQzEF0Ovsj7_TPsBJGkVJ-OiZsTXCe7EAmG5gNGGPBE5Gu14Rwb-eZ5r9RCAaPfhxR1yHYTAvCrku_6i2os7RLpT6UockKtX4QQSH2CMveNwqd6LdwhV8USZrczB2VYkAImngJC745-EWek1sVExYkqheGvC3J8O7D9H4JtaKD2zaq0rJzsIU0zb_wwax5-La-uRuPYvTXlO8B8IK4jjNMCQ',
  e: 'AQAB',
};

export const mockJwks = {
  keys: [mockRsaJwk],
};
