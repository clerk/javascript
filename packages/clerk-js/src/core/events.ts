import type { TokenResource } from '@clerk/types';

export const events = {
  TokenUpdate: 'token:update',
} as const;

type ClerkEvent = typeof events[keyof typeof events];
type EventType<T> = Record<string, unknown> & T;

type TokenUpdatePayload = { token: TokenResource | null };

type EventPayload = {
  [events.TokenUpdate]: EventType<TokenUpdatePayload>;
};

class ClerkEventBus extends EventTarget {
  #handlers = {} as Record<ClerkEvent, Array<any>>;

  on<E extends ClerkEvent, P extends EventPayload[E]>(eventName: E, callback: (evt: P) => void): void {
    const handler = (e: Event) => {
      callback((e as CustomEvent).detail);
    };

    if (!this.#handlers[eventName]) {
      this.#handlers[eventName] = [];
    }
    this.#handlers[eventName].push({ callback, handler });

    this.addEventListener(eventName, handler);
  }

  dispatch<E extends ClerkEvent, P extends EventPayload[E]>(eventName: E, data: P): void {
    this.dispatchEvent(new CustomEvent(eventName, { detail: data }));
  }

  off<E extends ClerkEvent, P extends EventPayload[E]>(eventName: E, callback?: (evt: P) => void): void {
    if (!this.#handlers[eventName]) {
      return;
    }

    if (callback) {
      const { handler } = this.#handlers[eventName].find(({ callback }) => callback === callback) || {};
      this.removeEventListener(eventName, handler);
    } else {
      this.#handlers[eventName].forEach(({ handler }) => {
        this.removeEventListener(eventName, handler);
      });
    }
  }
}

export const eventBus = new ClerkEventBus();
