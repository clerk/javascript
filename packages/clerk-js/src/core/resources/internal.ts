export type { default as Clerk } from '../clerk';
export * from './Base';
export * from './UserSettings';
export * from './AuthConfig';
export * from './Client';
export * from './DeletedObject';
export * from './DisplayConfig';
export * from './EmailAddress';
export * from './Environment';
export * from './Error';
export * from './ExternalAccount';
export * from './IdentificationLink';
export * from './Image';
export * from './PhoneNumber';
export * from './Organization';
export * from './OrganizationInvitation';
export * from './OrganizationMembership';
export * from './SamlAccount';
export * from './Session';
export * from './SessionWithActivities';
export * from './SignIn';
export * from './SignUp';
export * from './Token';
export * from './TOTP';
export * from './User';
export * from './Verification';
export * from './Web3Wallet';
