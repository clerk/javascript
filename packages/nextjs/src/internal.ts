/**
 * These need to be explicitly listed. Do not use an * here.
 * If you do, app router will break.
 */
export { MultisessionAppSupport } from './client-boundary/controlComponents';
