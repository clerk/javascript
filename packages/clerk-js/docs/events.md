# How to add an event

Using the new [events](https://github.com/clerk/javascript/blob/main/packages/clerk-js/src/core/events.ts) module in ClerkJS we can add a new event by defining the following:

```typescript
// file: src/core/events.ts

export const events = {
  // ...
  NewEvent: 'namespace_or_resource:action_or_identifier',
} as const;
// ... after TokenUpdatePayload ...
type NewEventPayload = { something: NewEventType };

type EventPayload = {
  //...
  [events.NewEvent]: EventType<NewEventPayload>;
};
```

Listen to event:

```typescript
import { eventBus, events } from './core/events';

eventBus.on(events.NewEvent, (obj: NewEventPayload) => {
  // do something
});
```

Dispatch event:

```typescript
import { eventBus, events } from './core/events';

const obj: NewEventPayload = { something: ... };
eventBus.dispatch(events.NewEvent, obj);
```

Disable events (if needed):

```typescript
import { eventBus, events } from './core/events';

// all handlers of event
eventBus.off(events.NewEvent);

// specific handler of event
// const handler = (obj: NewEventPayload) => { /* do something */ };
// eventBus.on(events.NewEvent, handler);
eventBus.off(events.NewEvent, handler);
```

## References

[Revamp Cookie Syncing](https://www.notion.so/Revamp-Cookie-Syncing-e9573ff022b24caf8f5a92e6be168b4d)
