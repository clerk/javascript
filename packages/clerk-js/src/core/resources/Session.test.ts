import type { SessionJSON } from '@clerk/types';

import { eventBus } from '../events';
import { clerkMock, createUser } from '../test/fixtures';
import { BaseResource, Session } from './internal';

export const mockJwt =
  'eyJhbGciOiJSUzI1NiIsImtpZCI6Imluc18yR0lvUWhiVXB5MGhYN0IyY1ZrdVRNaW5Yb0QiLCJ0eXAiOiJKV1QifQ.eyJhenAiOiJodHRwczovL2FjY291bnRzLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsImV4cCI6MTY2NjY0ODMxMCwiaWF0IjoxNjY2NjQ4MjUwLCJpc3MiOiJodHRwczovL2NsZXJrLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsIm5iZiI6MTY2NjY0ODI0MCwic2lkIjoic2Vzc18yR2JEQjRlbk5kQ2E1dlMxenBDM1h6Zzl0SzkiLCJzdWIiOiJ1c2VyXzJHSXBYT0VwVnlKdzUxcmtabjlLbW5jNlN4ciJ9.n1Usc-DLDftqA0Xb-_2w8IGs4yjCmwc5RngwbSRvwevuZOIuRoeHmE2sgCdEvjfJEa7ewL6EVGVcM557TWPW--g_J1XQPwBy8tXfz7-S73CEuyRFiR97L2AHRdvRtvGtwR-o6l8aHaFxtlmfWbQXfg4kFJz2UGe9afmh3U9-f_4JOZ5fa3mI98UMy1-bo20vjXeWQ9aGrqaxHQxjnzzC-1Kpi5LdPvhQ16H0dPB8MHRTSM5TAuLKTpPV7wqixmbtcc2-0k6b9FKYZNqRVTaIyV-lifZloBvdzlfOF8nW1VVH_fx-iW5Q3hovHFcJIULHEC1kcAYTubbxzpgeVQepGg';

describe('Session', () => {
  describe('getToken()', () => {
    let dispatchSpy;

    beforeEach(() => {
      dispatchSpy = jest.spyOn(eventBus, 'dispatch');
      BaseResource.clerk = clerkMock() as any;
    });

    afterEach(() => {
      dispatchSpy?.mockRestore();
      BaseResource.clerk = null as any;
    });

    it('dispatches token:update event on getToken', async () => {
      const session = new Session({
        status: 'active',
        id: 'session_1',
        object: 'session',
        user: createUser({}),
        last_active_organization_id: 'activeOrganization',
        actor: null,
        created_at: new Date().getTime(),
        updated_at: new Date().getTime(),
      } as SessionJSON);

      await session.getToken();

      expect(dispatchSpy).toBeCalledTimes(1);
      expect(dispatchSpy.mock.calls[0]).toMatchSnapshot();
    });

    it('dispatches token:update event on initilization with lastActiveToken', () => {
      new Session({
        status: 'active',
        id: 'session_1',
        object: 'session',
        user: createUser({}),
        last_active_organization_id: 'activeOrganization',
        last_active_token: { object: 'token', jwt: mockJwt },
        actor: null,
        created_at: new Date().getTime(),
        updated_at: new Date().getTime(),
      } as SessionJSON);

      expect(dispatchSpy).toBeCalledTimes(1);
      expect(dispatchSpy.mock.calls[0]).toMatchSnapshot();
    });
  });
});
