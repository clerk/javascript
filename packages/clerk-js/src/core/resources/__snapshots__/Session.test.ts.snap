// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Session getToken() dispatches token:update event on getToken 1`] = `
[
  "token:update",
  {
    "token": Token {
      "getRawString": [Function],
      "jwt": {
        "claims": {
          "__raw": "eyJhbGciOiJSUzI1NiIsImtpZCI6Imluc18yR0lvUWhiVXB5MGhYN0IyY1ZrdVRNaW5Yb0QiLCJ0eXAiOiJKV1QifQ.eyJhenAiOiJodHRwczovL2FjY291bnRzLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsImV4cCI6MTY2NjY0ODMxMCwiaWF0IjoxNjY2NjQ4MjUwLCJpc3MiOiJodHRwczovL2NsZXJrLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsIm5iZiI6MTY2NjY0ODI0MCwic2lkIjoic2Vzc18yR2JEQjRlbk5kQ2E1dlMxenBDM1h6Zzl0SzkiLCJzdWIiOiJ1c2VyXzJHSXBYT0VwVnlKdzUxcmtabjlLbW5jNlN4ciJ9.n1Usc-DLDftqA0Xb-_2w8IGs4yjCmwc5RngwbSRvwevuZOIuRoeHmE2sgCdEvjfJEa7ewL6EVGVcM557TWPW--g_J1XQPwBy8tXfz7-S73CEuyRFiR97L2AHRdvRtvGtwR-o6l8aHaFxtlmfWbQXfg4kFJz2UGe9afmh3U9-f_4JOZ5fa3mI98UMy1-bo20vjXeWQ9aGrqaxHQxjnzzC-1Kpi5LdPvhQ16H0dPB8MHRTSM5TAuLKTpPV7wqixmbtcc2-0k6b9FKYZNqRVTaIyV-lifZloBvdzlfOF8nW1VVH_fx-iW5Q3hovHFcJIULHEC1kcAYTubbxzpgeVQepGg",
          "azp": "https://accounts.inspired.puma-74.lcl.dev",
          "exp": 1666648310,
          "iat": 1666648250,
          "iss": "https://clerk.inspired.puma-74.lcl.dev",
          "nbf": 1666648240,
          "sid": "sess_2GbDB4enNdCa5vS1zpC3Xzg9tK9",
          "sub": "user_2GIpXOEpVyJw51rkZn9Kmnc6Sxr",
        },
        "encoded": {
          "header": "eyJhbGciOiJSUzI1NiIsImtpZCI6Imluc18yR0lvUWhiVXB5MGhYN0IyY1ZrdVRNaW5Yb0QiLCJ0eXAiOiJKV1QifQ",
          "payload": "eyJhenAiOiJodHRwczovL2FjY291bnRzLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsImV4cCI6MTY2NjY0ODMxMCwiaWF0IjoxNjY2NjQ4MjUwLCJpc3MiOiJodHRwczovL2NsZXJrLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsIm5iZiI6MTY2NjY0ODI0MCwic2lkIjoic2Vzc18yR2JEQjRlbk5kQ2E1dlMxenBDM1h6Zzl0SzkiLCJzdWIiOiJ1c2VyXzJHSXBYT0VwVnlKdzUxcmtabjlLbW5jNlN4ciJ9",
          "signature": "n1Usc-DLDftqA0Xb-_2w8IGs4yjCmwc5RngwbSRvwevuZOIuRoeHmE2sgCdEvjfJEa7ewL6EVGVcM557TWPW--g_J1XQPwBy8tXfz7-S73CEuyRFiR97L2AHRdvRtvGtwR-o6l8aHaFxtlmfWbQXfg4kFJz2UGe9afmh3U9-f_4JOZ5fa3mI98UMy1-bo20vjXeWQ9aGrqaxHQxjnzzC-1Kpi5LdPvhQ16H0dPB8MHRTSM5TAuLKTpPV7wqixmbtcc2-0k6b9FKYZNqRVTaIyV-lifZloBvdzlfOF8nW1VVH_fx-iW5Q3hovHFcJIULHEC1kcAYTubbxzpgeVQepGg",
        },
        "header": {
          "alg": "RS256",
          "kid": "ins_2GIoQhbUpy0hX7B2cVkuTMinXoD",
          "typ": "JWT",
        },
      },
      "pathRoot": "/client/sessions/session_1/tokens",
    },
  },
]
`;

exports[`Session getToken() dispatches token:update event on initilization with lastActiveToken 1`] = `
[
  "token:update",
  {
    "token": Token {
      "getRawString": [Function],
      "jwt": {
        "claims": {
          "__raw": "eyJhbGciOiJSUzI1NiIsImtpZCI6Imluc18yR0lvUWhiVXB5MGhYN0IyY1ZrdVRNaW5Yb0QiLCJ0eXAiOiJKV1QifQ.eyJhenAiOiJodHRwczovL2FjY291bnRzLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsImV4cCI6MTY2NjY0ODMxMCwiaWF0IjoxNjY2NjQ4MjUwLCJpc3MiOiJodHRwczovL2NsZXJrLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsIm5iZiI6MTY2NjY0ODI0MCwic2lkIjoic2Vzc18yR2JEQjRlbk5kQ2E1dlMxenBDM1h6Zzl0SzkiLCJzdWIiOiJ1c2VyXzJHSXBYT0VwVnlKdzUxcmtabjlLbW5jNlN4ciJ9.n1Usc-DLDftqA0Xb-_2w8IGs4yjCmwc5RngwbSRvwevuZOIuRoeHmE2sgCdEvjfJEa7ewL6EVGVcM557TWPW--g_J1XQPwBy8tXfz7-S73CEuyRFiR97L2AHRdvRtvGtwR-o6l8aHaFxtlmfWbQXfg4kFJz2UGe9afmh3U9-f_4JOZ5fa3mI98UMy1-bo20vjXeWQ9aGrqaxHQxjnzzC-1Kpi5LdPvhQ16H0dPB8MHRTSM5TAuLKTpPV7wqixmbtcc2-0k6b9FKYZNqRVTaIyV-lifZloBvdzlfOF8nW1VVH_fx-iW5Q3hovHFcJIULHEC1kcAYTubbxzpgeVQepGg",
          "azp": "https://accounts.inspired.puma-74.lcl.dev",
          "exp": 1666648310,
          "iat": 1666648250,
          "iss": "https://clerk.inspired.puma-74.lcl.dev",
          "nbf": 1666648240,
          "sid": "sess_2GbDB4enNdCa5vS1zpC3Xzg9tK9",
          "sub": "user_2GIpXOEpVyJw51rkZn9Kmnc6Sxr",
        },
        "encoded": {
          "header": "eyJhbGciOiJSUzI1NiIsImtpZCI6Imluc18yR0lvUWhiVXB5MGhYN0IyY1ZrdVRNaW5Yb0QiLCJ0eXAiOiJKV1QifQ",
          "payload": "eyJhenAiOiJodHRwczovL2FjY291bnRzLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsImV4cCI6MTY2NjY0ODMxMCwiaWF0IjoxNjY2NjQ4MjUwLCJpc3MiOiJodHRwczovL2NsZXJrLmluc3BpcmVkLnB1bWEtNzQubGNsLmRldiIsIm5iZiI6MTY2NjY0ODI0MCwic2lkIjoic2Vzc18yR2JEQjRlbk5kQ2E1dlMxenBDM1h6Zzl0SzkiLCJzdWIiOiJ1c2VyXzJHSXBYT0VwVnlKdzUxcmtabjlLbW5jNlN4ciJ9",
          "signature": "n1Usc-DLDftqA0Xb-_2w8IGs4yjCmwc5RngwbSRvwevuZOIuRoeHmE2sgCdEvjfJEa7ewL6EVGVcM557TWPW--g_J1XQPwBy8tXfz7-S73CEuyRFiR97L2AHRdvRtvGtwR-o6l8aHaFxtlmfWbQXfg4kFJz2UGe9afmh3U9-f_4JOZ5fa3mI98UMy1-bo20vjXeWQ9aGrqaxHQxjnzzC-1Kpi5LdPvhQ16H0dPB8MHRTSM5TAuLKTpPV7wqixmbtcc2-0k6b9FKYZNqRVTaIyV-lifZloBvdzlfOF8nW1VVH_fx-iW5Q3hovHFcJIULHEC1kcAYTubbxzpgeVQepGg",
        },
        "header": {
          "alg": "RS256",
          "kid": "ins_2GIoQhbUpy0hX7B2cVkuTMinXoD",
          "typ": "JWT",
        },
      },
      "pathRoot": "tokens",
    },
  },
]
`;
